ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/LagReportingAgent.java

```diff
@@ -0,0 +1,380 @@
+      final Set<HostInfo> hosts, final LagInfoKey lagInfoKey, final int partition) {
+    ImmutableMap.Builder<HostInfo, LagInfoEntity> builder = ImmutableMap.builder();
+    for (HostInfo host : hosts) {
+      LagInfoEntity lagInfo = receivedLagInfo.getOrDefault(host, EMPTY_HOST_LAG_INFO).getLagInfo()
+          .getOrDefault(lagInfoKey, Collections.emptyMap())
+          .getOrDefault(partition, null);
+      if (aliveHosts.containsKey(host) && lagInfo != null) {
+        builder.put(host, lagInfo);
+      }
+    }
+    return builder.build();
+  }
+
+  /**
+   * Returns a map of storeName -> partition -> LagInfoEntity.  Meant for being exposed in testing
+   * and debug resources.
+   */
+  public Map<String, Map<String, Map<Integer, LagInfoEntity>>> listAllLags() {
+      return receivedLagInfo.entrySet().stream()
+          .map(hostEntry -> Pair.of(
+              hostEntry.getKey().host() + ":" + hostEntry.getKey().port(),
+              hostEntry.getValue().getLagInfo().entrySet().stream()
+                  .map(storeNameEntry -> Pair.<String, Map<Integer, LagInfoEntity>>of(
+                      storeNameEntry.getKey().toString(),
+                      storeNameEntry.getValue().entrySet().stream()
+                          .map(partitionEntry -> Pair.of(
+                                partitionEntry.getKey(), partitionEntry.getValue()))
+                          .collect(ImmutableSortedMap.toImmutableSortedMap(
+                              Integer::compare, Pair::getLeft, Pair::getRight))))
+                  .collect(ImmutableSortedMap.toImmutableSortedMap(
+                      Comparator.comparing(String::toString), Pair::getLeft, Pair::getRight))))
+          .collect(ImmutableSortedMap.toImmutableSortedMap(
+              Comparator.comparing(String::toString), Pair::getLeft, Pair::getRight));
+  }
+
+  @Override
+  public void onHostStatusUpdated(final Map<String, HostStatusEntity> hostsStatusMap) {
+    aliveHosts.clear();
+    aliveHosts.putAll(hostsStatusMap.values().stream()
+        .filter(HostStatusEntity::getHostAlive)
+        .map(HostStatusEntity::getHostInfoEntity)
+        .map(hostInfoEntity -> new HostInfo(hostInfoEntity.getHost(), hostInfoEntity.getPort()))
+        .collect(Collectors.toMap(Function.identity(), hi -> true)));
+  }
+
+  /**
+   * Broadcast lags to remote hosts.
+   *
+   * <p>This is an asynchronous RPC and we do not handle the response returned from the remote
+   * server.</p>
+   */
+  class SendLagService extends AbstractScheduledService {
+
+    @Override
+    protected void runOneIteration() {
+      final List<PersistentQueryMetadata> currentQueries = engine.getPersistentQueries();
+      if (currentQueries.isEmpty()) {
+        return;
+      }
+
+      final Map<String, Map<Integer, LagInfo>> storeToPartitionToLagMap = currentQueries.stream()
+          .filter(Objects::nonNull)
+          .flatMap(pmd -> pmd.getStoreToPartitionToLagMap().entrySet().stream()
```

Comment:
could we just access the `KafkaStreams` instance here and fetch the lag map on this class, instead of QueryMetadata? Technically, the lag is not metadata about the query itself and it might be best to consolidate all the lag related code, to the agent

Reply:
Sure.  Moved the code that fetches the map to this class.